### ISSUE TITLE
Multi-GPU doesn't work for model(inputs) nor when computing the gradients

### ISSUE THREAD

**Issue by miguelCalado** [tensorflow/tensorflow#50593]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: Desktop
- TensorFlow installed from (source or binary): Source
- TensorFlow version (use command below): 2.4.0
- Python version: 3.8.5
- Bazel version (if compiling from source): 4.0.0
- GCC/Compiler version (if compiling from source): 9.3.0
- CUDA/cuDNN version: 11.0/8
- GPU model and memory: RTX5000-16GB

**Describe the current behavior**

When using multiple GPUs to perform inference on a model (e.g. the call method: `model(inputs)`) and calculate its gradients, the machine only uses one GPU, leaving the rest idle. 

**Describe the expected behavior**

I feel this should be the same as using `model.predict()` where all the GPUs are working. 

**Standalone code to reproduce the issue**

```
import tensorflow as tf
import numpy as np
import os

# Make the tf-data
path_filename_records = 'your_path_to_records'
bs = 128

dataset = tf.data.TFRecordDataset(path_filename_records)
dataset = (dataset
           .map(parse_record, num_parallel_calls=tf.data.experimental.AUTOTUNE)
           .batch(bs)
           .prefetch(tf.data.experimental.AUTOTUNE)
          )

# Load model trained using MirroredStrategy
path_to_resnet = 'your_path_to_resnet'
mirrored_strategy = tf.distribute.MirroredStrategy()
with mirrored_strategy.scope():
    resnet50 = tf.keras.models.load_model(path_to_resnet)

for pre_images, true_label in dataset:
    with tf.GradientTape() as tape:
       tape.watch(pre_images)
       outputs = resnet50(pre_images)
       grads = tape.gradient(outputs, pre_images)
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

You can profile the behavior of the GPUs with nvidia-smi. I don't know if it is supposed to be like this, both the `model(inputs)` and `tape.gradient` to not have multi-GPU support. But if it is, then it's a big problem because if you have a large dataset and need to calculate the gradients with respect to the inputs (e.g. interpretability porpuses) it might take days with one GPU. 
Another thing I tried was using `model.predict()` but this isn't possible with `tf.GradientTape`.

**Comment by tilakrayal** [comment]
@miguelCalado ,

TF v2.4 are built and tested against CUDA 11.0. Could you please check if you are facing the same error with CUDA 11.0 and cuDNN 8 and please let us know if you are facing same issue again.Please refer this link for [tested built configurations](https://www.tensorflow.org/install/source#gpu).Thanks!

**Comment by miguelCalado** [comment]
Yes, it still does. I edited the *system information* to not cause any more confusion.

**Comment by rmothukuru** [comment]
@miguelCalado,
Can you please try specifying the **`GPU Devices`** in the **`Mirrored Strategy`** as shown in the [Documentation of Mirrored Strategy](https://www.tensorflow.org/guide/distributed_training#mirroredstrategy), like:

**`mirrored_strategy = tf.distribute.MirroredStrategy(devices=["/gpu:0", "/gpu:1"])`**

and let us know how it goes? Thanks!

**Comment by miguelCalado** [comment]
Yes, the behavior still stands!

**Comment by miguelCalado** [comment]
Hi, any more updates on this issue? 

Here's the list of things I've tried:
1. Put all the code inside mirrored strategy scope.
2. Used different GPUs: I've tried A100, A6000 and RTX5000. Also changed the number of graphic cards and varied the batch size.
3. Specified a list of GPUs, for instance, `strategy = tf.distribute.MirroredStrategy(['/gpu:0', '/gpu:1'])`.
4. Added this `strategy = tf.distribute.MirroredStrategy(cross_device_ops=tf.distribute.HierarchicalCopyAllReduce())` as @Kaveh suggested.

**How do I know that only one GPU is working?**

I used the command `watch -n 1 nvidia-smi` in the terminal and observed that only one GPU is a 100%, the rest are at 0%.

**Comment by miguelCalado** [comment]
Here's a working example of it. I ask to whoever has more than one GPU to try and check if the issue still remains.

**Notebook**: [Working Example.ipynb](https://drive.google.com/file/d/1CoYVckmEQXp2Wf_PRlGtnrlGnF8smxvt/view?usp=sharing)

**Saved Model**: 
- [HDF5](https://drive.google.com/file/d/1Y0-fQytVsnHPs8JL6kKJr3tEL0mKNtjJ/view?usp=sharing)
- [Saved Format](https://drive.google.com/file/d/19oSIaUTtEy1q6rlDj8GzuWIvwAq_7XMi/view?usp=sharing)

**Comment by miguelCalado** [comment]
Problem solved! https://stackoverflow.com/a/68404933/8527630
Feel free to close this issue.
Thanks for all the help.
